kotlinx-coroutines-core/common/src/flow/operators/Merge.kt

```diff
@@ -89,15 +89,50 @@ public fun <T> Flow<Flow<T>>.flattenConcat(): Flow<T> = flow {
 @FlowPreview
 public fun <T> Flow<Flow<T>>.flattenMerge(concurrency: Int = 16, bufferSize: Int = 16): Flow<T> = flatMapMerge(concurrency, bufferSize) { it }
 
+/**
+ * Returns a flow that switches to the new flow produced by [transform] function every time the original flow emits a value.
+ * When switch on the newer flow is performed, the previous one is cancelled.
+ *
+ * For example, the following flow:
+ * ```
+ * flow {
+ *     emit("a")
+ *     delay(100)
+ *     emit("b")
+ * }.switchMap { value ->
+ *     flow {
+ *         emit(value + value)
+ *         delay(200)
+ *         emit(value + "_last")
+ *     }
+ * }
+ * ```
+ * will produce `aa bb b_last`
+ */
+@FlowPreview
+public fun <T, R> Flow<T>.switchMap(transform: suspend (value: T) -> Flow<R>): Flow<R> = flow {
+    coroutineScope {
+        var previousFlow: Job? = null
+        collect { value ->
+            // Linearize calls to emit as alternative to the channel. Bonus points for never-overlapping channels.
+            previousFlow?.cancelAndJoin()
```

Comment:
Is joining the previousFlow necessary when doing a switch-map? Wouldn't calling just `previousFlow?.cancel()` be enough?

Reply:
I'd rather say is necessary otherwise the flow that does not support cancellation (or e.g. does something like `finally { withContext(NonCancellable) { emit(theVeryLastItem) } }` will ovelap with a newer flow.

Reply:
True! 
But not joining doesn't affect the result and the previous 'collecting' of the flow is still canceled, I think. 
Would joining not 'slow down' the resulting stream?

Reply:
The joining will slow down the resulting stream _for good_, because we always wait for previous resources to be released, so the thing like `swithMap { doExpensiveRemoteCall() }` does not risk running us out of expensive resources.

Reply:
That is true. But is it up to the`switchMap` implementation to guard against this?

I think that if `doExpsensiveRemoteCall()` is so expensive, the `doExpensiveRemoteCall()` should guard against that issue (or wrap something around the call to guard against it).

In Rx, switchMap cancels/disposes the previous subscription. It doesn't wait for the subscription to be cancelled/disposed has completed.

Reply:
That is not true. RxJava calls `cancel()` (which invokes `dispose()`) on the active subscription synchronously. All work required to stop the previous stream will happen prior to the mapper being called for the new item and thus prior to the subscription to the returned stream.

https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableSwitchMap.java#L104-L107

Reply:
> I think that if `doExpsensiveRemoteCall()` is so expensive, the `doExpensiveRemoteCall()` should guard against that issue

That's exactly the point. `doExpsensiveRemoteCall()` _is guarded_, because it is coded like:
```
suspend funn doExpsensiveRemoteCall() {
    val con = openExpensiveConnection()
    try {
        // do call
    } finally {
        con.close() // Guarding against leak of expensive resource
    }
} 
```

So, now it is responsibility to of `switchMap` implementation to wait until `doExpsensiveRemoteCall` is complete (and releases all its resources)

Reply:
Then I stand corrected 😄 
When looking at the Rx docks, I thought that cancelation was done asynchronously:
http://reactivex.io/RxJava/javadoc/rx/Observable.html#switchMap-rx.functions.Func1-

Just out of curiosity:
Is it synchronous because the call to `inner.cancel()` is synchronous?
Or is it because the for-loop keeps checking until the inner has been canceled:
https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableSwitchMap.java#L121-L130

Reply:
It's the `inner.cancel()` which calls `dispose()` on the current upstream. That loop guards against the downstream canceling concurrently with an update from upstream.